File: src/LeetCode155.java
class MinStack {

    private Deque<Integer> minStack;
    private Deque<Integer> dataStack;

    public MinStack() {
        minStack = new ArrayDeque<>();
        dataStack = new ArrayDeque<>();
    }

    public void push(int x) {
        dataStack.push(x);
        minStack.push((minStack.isEmpty() || x < minStack.peek()) ?
                      x : minStack.peek());
    }

    public void pop() {
        dataStack.pop();
        minStack.pop();
    }

    public int top() {
        return dataStack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }
}



class MinStack {

    private Deque<Integer> minStack;
    private Deque<Integer> dataStack;

    public MinStack() {
        minStack = new ArrayDeque<>();
        dataStack = new ArrayDeque<>();
    }

    public void push(int x) {
        dataStack.push(x);
        if (minStack.isEmpty() || x <= minStack.peek()) {
            minStack.push(x);
        }
    }

    public void pop() {
        if (dataStack.peek().equals(minStack.peek())) {
            minStack.pop();
        }
        dataStack.pop();

    }

    public int top() {
        return dataStack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }
}

File: src/LeetCode155.cpp
class MinStack {
public:
    stack<int> dataStack;
    stack<int> minStack;
    MinStack() {

    }

    void push(int x) {
        dataStack.push(x);
        minStack.push((minStack.empty() || x < minStack.top()) ?
                      x : minStack.top());
    }

    void pop() {
        dataStack.pop();
        minStack.pop();
    }

    int top() {
        return dataStack.top();
    }

    int getMin() {
        return minStack.top();
    }
};
